enhancements/sig-architecture/32-extensiblescheduling/32-extensiblescheduling.md

```diff
@@ -0,0 +1,462 @@
+	ScoreName string `json:"scoreName"`
+}
+```
+
+### Let placement prioritizer support rating clusters with the customized score of the CR.
+How to maintain the lifecycle of AddOnPlacementScore CRs is out of scope in this proposal. Here we suppose there already are AddOnPlacementScore CRs created in each managed cluster namespace.
+A valid AddOnPlacementScore CR resembles as below:
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: AddOnPlacementScore
+metadata:
+  name: default
+  namespace: cluster1
+status:
+  conditions:
+  - lastTransitionTime: "2021-10-28T08:31:39Z"
+    message: AddOnPlacementScore updated successfully
+    reason: AddOnPlacementScoreUpdated
+    status: "True"
+    type: AddOnPlacementScoreUpdated
+  validUntil: "2021-10-29T18:31:39Z"
+  scores:
+  - name: "cpuratio"
+    value: 88
+  - name: "memratio"
+    value: 77
+  - name: "cpu"
+    value: 66
+  - name: "mem"
+    value: 55
+```
+The `AddOnPlacementScore` resource name is "default" in this example, the namespace "cluster1" is the managed cluster namespace.
+The `status.scores` contains a list of score items, for example, score "cpuratio" has value "88".
+
+As an end user, can define a placement as below to use the scores in `AddOnPlacementScore` to sort clusters.
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: Placement
+metadata:
+  name: placement
+  namespace: ns1
+spec:
+  numberOfClusters: 3
+  prioritizerPolicy:
+    mode: Exact
+    configurations:
+      - scoreCoordinate:
+          type: BuildIn
+          buildIn: Steady
+        weight: 3
+      - scoreCoordinate:
+          type: AddOn
+          addOn:
+            resourceName: default
+            scoreName: cpuratio
+        weight: 1
+```
+
+In the above example, the placement defines buildin prioritizer "Steady" with weight 3, and as well as using score "cpuratio" in "default" AddOnPlacementScore to sort clusters.
+
+When this placement.yaml is created, the placement controller sorts the clusters with "BuildIn" prioritizer "Steady" and "AddOn" score "cpuratio". The "BuildIn" prioritizer scheduling behavior keeps the same before, the "AddOn" type prioritizer scheduling process is as below:
+
+- First schedule
+
+The prioritizer will get customized score "cpuratio" from "default" `AddOnPlacementScore`, and calculate a final score for each managed clusters.
+
+- Reschedule
+
+The scores inside `AddOnPlacementScore` are supposed to be updated frequently. Rather than watch the changes in `AddOnPlacementScore` and trigger reschedule, the scheduler framework would periodically (eg. every 5 minutes) reschedule the placement.
+
+**What if no valid score is generated?**
+
+What if some of the scores are invalid? For example, the AddOnPlacementScore CR is not created for some managed clusters, the score is missing in some CR, the score is expired. To handle the invalid cases, we add a PreScore() in the Prioritizer interface, to do some pre-check before Score().
```

Comment:
Why is it desireable to effectively suspend a placement because a particular addonplacementscore isn't present?  This seems to produce a situation where a placement ceases to prioritize those things that it can in the absence of complete information.  Is there a discussion of the tradeoff and why this is good?

Reply:
The original consideration is AddOnPlacementScore CRs are created on-demand, so don't want to make decisions when most of the scores are not ready. 

As you mentioned, it's reasonably expect that the AddOnPlacementScoring controller is created before the placement attempts to use it. It makes sense to give a default score 0 if it isn't available.